### PR TITLE
feat: section loader

### PR DESCRIPTION
**Pull request by davidnuescheler**
Test URLs:
- Before: https://main--aem-boilerplate--adobe.hlx.live/
- After: https://section-loader--aem-boilerplate--adobe.hlx.live/

this is just a test, do not merge.

### CODE REVIEW

**Comment by davidnuescheler** [comment]
this removes complexity and makes LCP guessing more flexible.

**Comment by davidnuescheler** [comment]
image

**Comment by davidnuescheler** [comment]
image

File: scripts/aem.js
Comment (shsteimer) [inline scripts/aem.js:704]: @davidnuescheler this may be counter-intuitive, but if we go this route, would it make sense to actually load the blocks within a section in reverse order.

My thinking:

I've had projects where you end up with nested blocks via auto-blocking. when these blocks are loaded sequentially, the outer block is loaded before the inner block, meaning that the outer block decorate function cannot make decisions based on the content of the inner blocks (because it isn't yet fully formed). There are ways around this on a project basis of course, but it can get messy.

we'd never want to load the blocks for a full page in reverse order as it would delay appearance of top of the page, but since no section gets reveled until all of it's blocks are loaded, maybe within each section it would be safe/performant to do so, allowing nested blocks to be loaded before their `"parent"` blocks, and thus resolving the issue.

(or maybe this is only a problem on a small enough percentage of projects to just not worry about it. feel free to just ignore this)

Comment (davidnuescheler) [reply scripts/aem.js:704]: that's very interesting indeed, i definitely left nested blocks via auto blocking kind of out of the equation... i think you are right that reverse order would yield better results in the sense that it may simplify that (special) case... TBH though, i have not been in a situation where i needed to make decisions in a block based on the nested block (in JS that is).
as you point out, i agree that it would be counter intuitive, and based on the sequence of https://github.com/adobe/helix-home/blob/main/manifesto.md#intuitive-simple--consistent-p i would probably at this stage defer to a later stage to incorporate this. personally, i think the nesting probably should be addressed with a more recursive approach at some point... at least that would be most intuitive to me.

File: scripts/aem.js
Comment (auniverseaway) [inline scripts/aem.js:721]:
```
for (const [idx, section] of sections.entries()) {
  await loadSection(section);
}
```

**Comment by auniverseaway** [comment]
I'm just seeing this. I think this PR is great. This is much closer to what we did on Milo.

https://github.com/da-sites/nexter/blob/main/nx/scripts/nexter.js#L125-L136

One thing we also did in Milo is that we do not allow a block to "pollute up" their class names into sections. We saw a lot of CSS written in Express due to competing section styles because of block names being attached on sections.

Screenshot 2024-03-16 at 1 23 32 PM

**Our guiding principles on this topic:**

1. A block should never change the behavior of a section.
2. A section should never impact the internals of a block.

File: scripts/aem.js
Comment (kptdobe) [inline scripts/aem.js:707]:
```suggestion
    const promises = blocks.map(block => loadBlock(block));
    await Promise.all(promises);
```
The `for` loop still makes the block loading sequential, i.e. block1, block2... blockN. If block 1 requires something remote and slow, it "blocks" all the other blocks processing. This can be avoided by parallelizing the block loading sequence.
The section is hidden anyway, no visible impact.

**Review by kptdobe — CHANGES_REQUESTED** [review]
I have tested the loading sequence in the context of the bulk project (not in production -yet). It helps solving several issues, like the case where the LCP image is part of block 2: no way to make it work with current implementation.
I even think the current implementation (`waitForLCP`) is incorrect: we wait for the first image of the first block but it only gets displayed in the lazy phase when the first section is processed. We ned to change that asap.

File: scripts/aem.js
Comment (kptdobe) [reply scripts/aem.js:707]: It seems I have proven in the past that this does not bring any real performance benefit AND makes the block code potentially more complicated (for example, if 2 blocks uses the same resource - like metadata.json, you need a singleton to make sure it is loaded only once...). This makes this change unnecessary.

**Review by kptdobe — APPROVED** [review]

**Comment by kptdobe** [comment]
Working on the aem-lib version of this PR.

File: scripts/aem.js
Comment (kptdobe) [reply scripts/aem.js:721]: Linting does not like this approach...
image

**Comment by kptdobe** [comment]
I have created the PR in aem-lib to match all the changes required in aem.js: https://github.com/adobe/aem-lib/pull/80

I have aligned the aem.js of this PR with the build result of the aem-lib PR for testing purposes. Once the aem-lib PR is merged, this one can be merged without the aem.js changes.

**Comment by kptdobe** [comment]
Replaces https://github.com/adobe/aem-boilerplate/pull/382.

**Pull request merged**
